pkg/select/blockdevice/filters.go

```diff
@@ -0,0 +1,199 @@
+	filteredList := bdList
+	for _, key := range filterKeys {
+		filteredList = filterFuncMap[key](filteredList, c.ClaimSpec)
+		glog.Info("Key : ", key)
+		glog.Info("Items:", filteredList.Items)
+	}
+	return filteredList
+}
+
+// filterActive filters out active Blockdevices from the BDList
+func filterActive(originalBD *apis.BlockDeviceList, spec *apis.DeviceClaimSpec) *apis.BlockDeviceList {
+	filteredBDList := &apis.BlockDeviceList{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "BlockDevice",
+			APIVersion: "openebs.io/v1alpha1",
+		},
+	}
+
+	for _, bd := range originalBD.Items {
+		if bd.Status.State == controller.NDMActive {
+			filteredBDList.Items = append(filteredBDList.Items, bd)
+		}
+	}
+	return filteredBDList
+}
+
+// filterUnclaimed returns only unclaimed devices
+func filterUnclaimed(originalBD *apis.BlockDeviceList, spec *apis.DeviceClaimSpec) *apis.BlockDeviceList {
+	filteredBDList := &apis.BlockDeviceList{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "BlockDevice",
+			APIVersion: "openebs.io/v1alpha1",
+		},
+	}
+
+	for _, bd := range originalBD.Items {
+		if bd.Status.ClaimState == apis.BlockDeviceUnclaimed {
+			filteredBDList.Items = append(filteredBDList.Items, bd)
+		}
+	}
+	return filteredBDList
+}
+
+// filterDeviceType returns only BDs which match the device type
+func filterDeviceType(originalBD *apis.BlockDeviceList, spec *apis.DeviceClaimSpec) *apis.BlockDeviceList {
+
+	// if no device type is specified, filter will not be effective
+	if spec.DeviceType == "" {
+		return originalBD
+	}
+
+	filteredBDList := &apis.BlockDeviceList{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "BlockDevice",
+			APIVersion: "openebs.io/v1alpha1",
+		},
+	}
+
+	for _, bd := range originalBD.Items {
+		if bd.Spec.Details.DeviceType == spec.DeviceType {
+			filteredBDList.Items = append(filteredBDList.Items, bd)
+		}
+	}
+	return filteredBDList
+}
+
+// filterVolumeMode returns only BDs which matches the specified volume mode
+func filterVolumeMode(originalBD *apis.BlockDeviceList, spec *apis.DeviceClaimSpec) *apis.BlockDeviceList {
+
+	volumeMode := spec.Details.BlockVolumeMode
+
+	// if volume mode is not specified in claim spec, this filter will not be effective
+	if volumeMode == "" {
+		return originalBD
+	}
+
+	filteredBDList := &apis.BlockDeviceList{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "BlockDevice",
+			APIVersion: "openebs.io/v1alpha1",
+		},
+	}
+
+	for _, bd := range originalBD.Items {
+		switch volumeMode {
+		case apis.VolumeModeBlock:
+			// if blockvolume mode, FS and Mountpoint should be empty
+			if bd.Spec.FileSystem.Type != "" || bd.Spec.FileSystem.Mountpoint != "" {
+				continue
+			}
+
+		case apis.VolumeModeFileSystem:
+			// if FSVolume mode, either FSType should be blank or FStype should
+			// match with format given in claimspec
+			if bd.Spec.FileSystem.Type != "" && bd.Spec.FileSystem.Type != spec.Details.DeviceFormat {
```

Comment:
Just thinking aloud here:
```
bd.fsType != "" (is valid)
bdc.fsType != "" and bdc.fsType == bd.fsType (is allowed) 
```

So should the continue condition be : 
```
bd.fsType == ""  or 
bdc.fsType != "" and bdc.fsType != bd.fsType 
```

Reply:
Which means we should not select if
1) BD does not have a filesystem
or
2) BDC does not have a filesystem specified or the specified filesystem does not match with the FS in BD.